How many late days are you using for this assignment? 0

In [2]:
# Imports and pip installations (if needed)
%pip install pandas


Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.0 -> 25.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


# Part 1: Load the dataset

In [3]:
# Load the given datasets
import pandas as pd
df = pd.read_csv('./chronic_kidney_disease_categorical.csv')
df2 = pd.read_csv('chronic_kidney_disease_numerical.csv')

# Print the data
display(df,df2)


,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,224481,3.0,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,good,yes,no,ckd
1,992643,0.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd
2,308740,1.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,good,yes,no,ckd
3,450314,1.0,0.0,NaN,normal,notpresent,notpresent,no,no,no,good,no,yes,ckd
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
425,769512,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
426,872034,2.0,1.0,abnormal,abnormal,notpresent,present,yes,yes,yes,poor,no,no,ckd
427,643007,NaN,NaN,NaN,normal,notpresent,notpresent,yes,yes,yes,poor,no,no,ckd
428,840392,NaN,NaN,NaN,NaN,notpresent,notpresent,yes,no,no,poor,no,no,ckd


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,7300.0,6.4
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,7200.0,5.5
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,5500.0,5.7
3,764115,70.0,90.0,144.0,125.0,4.0,136.0,4.6,12.0,37.0,8200.0,4.5
4,240975,47.0,80.0,95.0,35.0,0.9,140.0,4.1,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...
415,522912,56.0,80.0,165.0,55.0,1.8,NaN,NaN,13.5,40.0,11800.0,5.0
416,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,7500.0,3.4
417,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,9400.0,6.4
418,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,12400.0,3.9


# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: The dataset is designed to predict if a person may have chronic kidney diseases based on some biological attributes. It includes 400 instances(rows) which means 400 patients or samples and 24 features (columns) which are the attributes and 1 class label to indicate whether the patient has the mentioned disease or not. There are some missing data which means some values will be null as they not exist and we have to process them before feed into any model. The features in the dataset are both numerical and categorical, with 11  numeric and 14  nominal included the label to be precise  Here are 3 examples each of how they look like:
- Numerical features:
    + Age: 20 (years)
    + Blood pressure: 80 (mm/Hg)
    + Red Blood Cell Count: 6.4 (millions/cmm)
- Categorical features:
    + Appetite: good
    + Anemia: yes
    + Coronary Artery Disease: No

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [4]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
check_duplicate_numerical = df2.duplicated()
print(f"Total number of duplicate rows in numerical dataset: {check_duplicate_numerical.sum()}")

# Drop these duplicate rows
no_duplicate_numerical = df2.drop_duplicates()
print("numerical after no duplication: ",no_duplicate_numerical.shape)

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
duplicate_rows_categorical = df.duplicated()
print(f"Number of duplicate rows in categorical dataset: {duplicate_rows_categorical.sum()}")
no_duplicate_categorical = df.drop_duplicates()
print("categorical after no duplication: ",no_duplicate_categorical.shape)



Total number of duplicate rows in numerical dataset: 30
numerical after no duplication:  (390, 12)
Number of duplicate rows in categorical dataset: 30
categorical after no duplication:  (400, 14)


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: I think the answer is all unique IDs are not present in both datasets. When I check the shapes of the datasets after removing duplicates, I observe that categorical dataset has 400 rows while numerical only has 390 rows. This means there are 10 unique IDs that exist in the categorical dataset but not in the numerical dataset. I feel this is because some patients may have had only categorical data reported but not numerical measurements. In addition there may be errors during data collection process, which leads to missing values. Moreover, it is possible that these 2 are collected in 2 different timeframes, and the person has the data collected for the first time for categorical data reports but they don't come back to finish process for numerical report. Last but not least, the data may have been processed before for some reasons and some valued get trimmed off. In the combined table, my assumption is that either those 10 rows will get removed when merging or they will all get included in the merged table but the missing fields will be replaced with some placeholder values like null/NaN/void/empty or something like that. 

In [5]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
combined_df = pd.merge(no_duplicate_categorical, no_duplicate_numerical, on='unique_id', how='inner', validate='one_to_one')

# Print the combined dataset
display(combined_df)


,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,...,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
0,224481,3.0,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,...,60.0,125.0,21.0,1.3,137.0,3.4,15.0,46.0,NaN,NaN
1,992643,0.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,...,70.0,82.0,36.0,1.1,150.0,3.5,14.5,52.0,9400.0,6.1
2,308740,1.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,...,70.0,139.0,29.0,1.0,NaN,NaN,10.5,32.0,NaN,NaN
3,450314,1.0,0.0,NaN,normal,notpresent,notpresent,no,no,no,...,NaN,NaN,16.0,0.7,138.0,3.2,8.1,NaN,NaN,NaN
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,70.0,75.0,31.0,1.2,141.0,3.5,16.5,54.0,7800.0,6.2
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
385,789906,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,70.0,83.0,42.0,1.2,139.0,3.7,16.2,50.0,9300.0,5.4
386,490253,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,80.0,81.0,46.0,0.6,135.0,4.9,13.9,48.0,6900.0,4.9
387,492993,3.0,0.0,NaN,normal,notpresent,notpresent,yes,no,no,...,80.0,76.0,162.0,9.6,141.0,4.9,7.6,24.0,3800.0,2.8
388,273038,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,60.0,117.0,22.0,1.2,138.0,3.5,13.0,45.0,5200.0,5.6


## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [6]:
# Calculate the percentage of rows that contain atleast one missing value
rows_with_missing = combined_df.isna().any(axis='columns')
percentage_missing = (rows_with_missing.sum() / combined_df.shape[0]) * 100

# Print %
print(f"Percentage of rows with at least one missing value: {percentage_missing:}%")

# Drop these rows from the dataset
cleaned_df = combined_df.dropna()

# Print the Dataset
display(cleaned_df)

Percentage of rows with at least one missing value: 60.76923076923077%


,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,...,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,70.0,75.0,31.0,1.2,141.0,3.5,16.5,54.0,7800.0,6.2
5,114717,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,80.0,122.0,25.0,0.8,138.0,5.0,17.1,41.0,9100.0,5.2
7,901528,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,80.0,118.0,44.0,0.7,137.0,3.5,14.8,45.0,9300.0,4.7
8,433613,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,80.0,97.0,40.0,0.6,150.0,4.5,14.2,48.0,10500.0,5.0
9,621332,1.0,0.0,abnormal,normal,notpresent,notpresent,no,yes,no,...,80.0,303.0,35.0,1.3,122.0,3.5,10.4,35.0,10900.0,4.3
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
382,996076,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,60.0,105.0,49.0,1.2,150.0,4.7,15.7,44.0,10400.0,6.2
384,397388,2.0,2.0,normal,normal,notpresent,present,yes,no,yes,...,100.0,280.0,35.0,3.2,143.0,3.5,13.0,40.0,9800.0,4.2
385,789906,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,70.0,83.0,42.0,1.2,139.0,3.7,16.2,50.0,9300.0,5.4
386,490253,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,80.0,81.0,46.0,0.6,135.0,4.9,13.9,48.0,6900.0,4.9


## Part 3.4: Sort the dataset according to the Labels

In [7]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
sorted_df = cleaned_df.sort_values(by='Target').reset_index(drop=True)
# Print the dataset
display(sorted_df)

,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,...,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
0,534245,2.0,0.0,normal,abnormal,present,notpresent,no,no,no,...,70.0,117.0,52.0,2.2,136.0,3.8,10.0,30.0,19100.0,3.7
1,349892,1.0,3.0,abnormal,abnormal,notpresent,notpresent,yes,yes,yes,...,70.0,424.0,55.0,1.7,138.0,4.5,12.6,37.0,10200.0,4.1
2,343710,2.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,yes,yes,...,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,9200.0,3.2
3,102677,3.0,0.0,normal,abnormal,notpresent,notpresent,yes,no,no,...,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,3800.0,2.1
4,514721,2.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,...,90.0,129.0,107.0,6.7,131.0,4.8,9.1,29.0,6400.0,3.4
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,489160,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,60.0,79.0,50.0,0.5,145.0,5.0,17.6,51.0,6500.0,5.0
149,530138,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,80.0,111.0,34.0,1.1,145.0,4.0,14.3,41.0,7200.0,5.0
150,261137,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,60.0,105.0,39.0,0.5,135.0,3.9,14.7,43.0,5800.0,6.2
151,322882,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,80.0,128.0,38.0,0.6,135.0,3.9,13.1,45.0,6200.0,4.5


## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [8]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
df_encoded = pd.get_dummies(sorted_df, columns=categorical_columns, dtype=int)

# Print the dataset
display(df_encoded)


,unique_id,al,su,age,bp,bgr,bu,sc,sod,pot,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,534245,2.0,0.0,45.0,70.0,117.0,52.0,2.2,136.0,3.8,...,1,0,1,0,1,0,1,0,1,0
1,349892,1.0,3.0,59.0,70.0,424.0,55.0,1.7,138.0,4.5,...,0,1,1,0,1,0,1,0,1,0
2,343710,2.0,0.0,61.0,80.0,173.0,148.0,3.9,135.0,5.2,...,0,1,0,1,0,1,0,1,1,0
3,102677,3.0,0.0,59.0,70.0,76.0,186.0,15.0,135.0,7.6,...,1,0,0,1,0,1,0,1,1,0
4,514721,2.0,0.0,56.0,90.0,129.0,107.0,6.7,131.0,4.8,...,1,0,1,0,1,0,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,489160,0.0,0.0,28.0,60.0,79.0,50.0,0.5,145.0,5.0,...,1,0,1,0,1,0,1,0,0,1
149,530138,0.0,0.0,23.0,80.0,111.0,34.0,1.1,145.0,4.0,...,1,0,1,0,1,0,1,0,0,1
150,261137,0.0,0.0,35.0,60.0,105.0,39.0,0.5,135.0,3.9,...,1,0,1,0,1,0,1,0,0,1
151,322882,0.0,0.0,33.0,80.0,128.0,38.0,0.6,135.0,3.9,...,1,0,1,0,1,0,1,0,0,1


In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: When using a single column for a binary variable, this would be equivalent in most downstream models. This is because model can still learn the same information and the only difference is the way of representation. However, for categorical variables with more than 2 values (like 5 types of condiments) or more, using a single column would not work because the model will misinterpret the order, and this will affect significantly to the performance and very likely gives wrong output. This is because the model misinterprets the information and it just like we use a different unexpected data to train the models. Also, some models may not able to process categorical type in a single column. 

In the sandwich rating prediction example, we can use one-hot encoding to tell the models the difference and independence of each condiment without accidentally establishing relationships (correlation) between them. Using a single column would make the model to think the condiments are ordinal while they are actually nominal, potentially missing the true connection between condiment type and sandwich rating, giving false results and bad predicting performance.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [9]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

mean = df_encoded[numerical_columns].mean()
std = df_encoded[numerical_columns].std()

lower_bound = mean - 3 * std
upper_bound = mean + 3 * std

mask = pd.DataFrame()
for col in numerical_columns:
    column_mask = (df_encoded[col] >= lower_bound[col]) & (df_encoded[col] <= upper_bound[col])
    mask[col] = column_mask

all_columns_mask = mask.all(axis='columns')
df_no_outliers = df_encoded[all_columns_mask]
print(f"Number of outliers removed: {df_encoded.shape[0] - df_no_outliers.shape[0]}")

# Print the dataset
display(df_no_outliers)

Number of outliers removed: 19


,unique_id,al,su,age,bp,bgr,bu,sc,sod,pot,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
2,343710,2.0,0.0,61.0,80.0,173.0,148.0,3.9,135.0,5.2,...,0,1,0,1,0,1,0,1,1,0
4,514721,2.0,0.0,56.0,90.0,129.0,107.0,6.7,131.0,4.8,...,1,0,1,0,1,0,1,0,1,0
7,124304,3.0,4.0,69.0,70.0,264.0,87.0,2.7,130.0,4.0,...,0,1,1,0,0,1,1,0,1,0
8,980291,3.0,0.0,62.0,70.0,122.0,42.0,1.7,136.0,4.7,...,1,0,1,0,1,0,1,0,1,0
10,615697,4.0,0.0,6.0,60.0,94.0,67.0,1.0,135.0,4.9,...,1,0,0,1,1,0,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,489160,0.0,0.0,28.0,60.0,79.0,50.0,0.5,145.0,5.0,...,1,0,1,0,1,0,1,0,0,1
149,530138,0.0,0.0,23.0,80.0,111.0,34.0,1.1,145.0,4.0,...,1,0,1,0,1,0,1,0,0,1
150,261137,0.0,0.0,35.0,60.0,105.0,39.0,0.5,135.0,3.9,...,1,0,1,0,1,0,1,0,0,1
151,322882,0.0,0.0,33.0,80.0,128.0,38.0,0.6,135.0,3.9,...,1,0,1,0,1,0,1,0,0,1


## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [10]:
# Normalize the all Numerical Attributes in the dataset.
min_values = df_no_outliers[numerical_columns].min()
max_values = df_no_outliers[numerical_columns].max()

df_normalized = df_no_outliers.copy()
for col in numerical_columns:
    df_normalized[col] = (df_no_outliers[col] - min_values[col]) / (max_values[col] - min_values[col])

# Print the dataset
display(df_normalized)

,unique_id,al,su,age,bp,bgr,bu,sc,sod,pot,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
2,343710,2.0,0.0,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.657143,...,0,1,0,1,0,1,0,1,1,0
4,514721,2.0,0.0,0.675676,0.75,0.253219,0.633987,0.777778,0.366667,0.542857,...,1,0,1,0,1,0,1,0,1,0
7,124304,3.0,4.0,0.851351,0.25,0.832618,0.503268,0.283951,0.333333,0.314286,...,0,1,1,0,0,1,1,0,1,0
8,980291,3.0,0.0,0.756757,0.25,0.223176,0.209150,0.160494,0.533333,0.514286,...,1,0,1,0,1,0,1,0,1,0
10,615697,4.0,0.0,0.000000,0.00,0.103004,0.372549,0.074074,0.500000,0.571429,...,1,0,0,1,1,0,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,489160,0.0,0.0,0.297297,0.00,0.038627,0.261438,0.012346,0.833333,0.600000,...,1,0,1,0,1,0,1,0,0,1
149,530138,0.0,0.0,0.229730,0.50,0.175966,0.156863,0.086420,0.833333,0.314286,...,1,0,1,0,1,0,1,0,0,1
150,261137,0.0,0.0,0.391892,0.00,0.150215,0.189542,0.012346,0.500000,0.285714,...,1,0,1,0,1,0,1,0,0,1
151,322882,0.0,0.0,0.364865,0.50,0.248927,0.183007,0.024691,0.500000,0.285714,...,1,0,1,0,1,0,1,0,0,1


## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: I observe that the column `unique_id` is redundant and may introduce unexpected behavior of our models. This column store unique identifiers for each patient, which has no predictive meaning with the label(or kidney disease status). Including identifier columns in learning models can lead to false correlation as the model may think there is a pattern and the ids are important parts of the data but actually they do not have any value in learning. So when the models predict, it may rely on the pattern of these arbitrary IDs rather than the medical features which are the real expected features we focus on. In addition, it will increase the complexity for the model unnecessarily as there are no purposes to present the IDs. In other words, this column is only useful for record identification, but adds no value for pattern recognition or prediction tasks. Therefore, it should be removed from the modeling dataset.

In [11]:
#Remove that column
final_df = df_normalized.drop(columns='unique_id')
# Print the dataset
display(final_df)

,al,su,age,bp,bgr,bu,sc,sod,pot,hemo,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
2,2.0,0.0,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.657143,0.172131,...,0,1,0,1,0,1,0,1,1,0
4,2.0,0.0,0.675676,0.75,0.253219,0.633987,0.777778,0.366667,0.542857,0.286885,...,1,0,1,0,1,0,1,0,1,0
7,3.0,4.0,0.851351,0.25,0.832618,0.503268,0.283951,0.333333,0.314286,0.565574,...,0,1,1,0,0,1,1,0,1,0
8,3.0,0.0,0.756757,0.25,0.223176,0.209150,0.160494,0.533333,0.514286,0.573770,...,1,0,1,0,1,0,1,0,1,0
10,4.0,0.0,0.000000,0.00,0.103004,0.372549,0.074074,0.500000,0.571429,0.352459,...,1,0,0,1,1,0,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,0.0,0.0,0.297297,0.00,0.038627,0.261438,0.012346,0.833333,0.600000,0.983607,...,1,0,1,0,1,0,1,0,0,1
149,0.0,0.0,0.229730,0.50,0.175966,0.156863,0.086420,0.833333,0.314286,0.713115,...,1,0,1,0,1,0,1,0,0,1
150,0.0,0.0,0.391892,0.00,0.150215,0.189542,0.012346,0.500000,0.285714,0.745902,...,1,0,1,0,1,0,1,0,0,1
151,0.0,0.0,0.364865,0.50,0.248927,0.183007,0.024691,0.500000,0.285714,0.614754,...,1,0,1,0,1,0,1,0,0,1


## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [12]:
# Export the dataframe to a new csv file
final_df.to_csv('output.csv')
# Export the dataframe to a new json file
final_df.to_json('output.json')

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [21]:
final_df.head(15)

,al,su,age,bp,bgr,bu,sc,sod,pot,hemo,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
2,2.0,0.0,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.657143,0.172131,...,0,1,0,1,0,1,0,1,1,0
4,2.0,0.0,0.675676,0.75,0.253219,0.633987,0.777778,0.366667,0.542857,0.286885,...,1,0,1,0,1,0,1,0,1,0
7,3.0,4.0,0.851351,0.25,0.832618,0.503268,0.283951,0.333333,0.314286,0.565574,...,0,1,1,0,0,1,1,0,1,0
8,3.0,0.0,0.756757,0.25,0.223176,0.209150,0.160494,0.533333,0.514286,0.573770,...,1,0,1,0,1,0,1,0,1,0
10,4.0,0.0,0.000000,0.00,0.103004,0.372549,0.074074,0.500000,0.571429,0.352459,...,1,0,0,1,1,0,1,0,1,0
11,3.0,1.0,0.662162,0.50,0.618026,0.411765,0.432099,0.566667,0.571429,0.434426,...,1,0,1,0,0,1,1,0,1,0
13,1.0,0.0,0.540541,0.00,0.399142,0.535948,0.358025,0.700000,0.314286,0.344262,...,1,0,1,0,1,0,1,0,1,0
14,4.0,2.0,0.783784,1.00,0.399142,0.287582,0.839506,0.666667,0.485714,0.188525,...,1,0,1,0,0,1,1,0,1,0
15,4.0,0.0,0.567568,0.50,0.107296,1.000000,0.901235,0.533333,0.257143,0.344262,...,1,0,1,0,1,0,0,1,1,0
16,4.0,3.0,0.851351,0.25,0.618026,0.562092,0.728395,0.000000,0.285714,0.311475,...,0,1,1,0,0,1,0,1,1,0


Paste here: 
| al  | su  | age       | bp   | bgr       | bu       | sc       | sod      | pot      | hemo     | ... | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|-----|-----|-----------|------|-----------|----------|----------|----------|----------|----------|-----|--------|---------|------------|------------|-------|--------|--------|---------|------------|----------------|
| 2   | 2.0 | 0.0      | 0.743 | 0.50      | 0.442060 | 0.901961 | 0.432099 | 0.500000 | 0.657143 | ... | 0      | 1       | 0          | 1          | 0     | 1      | 0      | 1       | 1          | 0              |
| 4   | 2.0 | 0.0      | 0.676 | 0.75      | 0.253219 | 0.633987 | 0.777778 | 0.366667 | 0.542857 | ... | 1      | 0       | 1          | 0          | 1     | 0      | 1      | 0       | 1          | 0              |
| 7   | 3.0 | 4.0      | 0.851 | 0.25      | 0.832618 | 0.503268 | 0.283951 | 0.333333 | 0.565574 | ... | 0      | 1       | 1          | 0          | 0     | 1      | 1      | 0       | 1          | 0              |
| 8   | 3.0 | 0.0      | 0.757 | 0.25      | 0.223176 | 0.209150 | 0.160494 | 0.533333 | 0.573770 | ... | 1      | 0       | 1          | 0          | 1     | 0      | 1      | 0       | 1          | 0              |
| 10  | 4.0 | 0.0      | 0.000 | 0.00      | 0.103004 | 0.372549 | 0.074074 | 0.500000 | 0.571429 | ... | 1      | 0       | 0          | 1          | 1     | 0      | 1      | 0       | 1          | 0              |
| 11  | 3.0 | 1.0      | 0.662 | 0.50      | 0.618026 | 0.411765 | 0.432099 | 0.566667 | 0.571429 | ... | 1      | 0       | 1          | 0          | 0     | 1      | 1      | 0       | 1          | 0              |
| 13  | 1.0 | 0.0      | 0.541 | 0.00      | 0.399142 | 0.535948 | 0.358025 | 0.700000 | 0.314286 | ... | 1      | 0       | 1          | 0          | 1     | 0      | 1      | 0       | 1          | 0              |
| 14  | 4.0 | 2.0      | 0.784 | 1.00      | 0.399142 | 0.287582 | 0.839506 | 0.666667 | 0.485714 | ... | 1      | 0       | 1          | 0          | 0     | 1      | 1      | 0       | 1          | 0              |
| 15  | 4.0 | 0.0      | 0.568 | 0.50      | 0.107296 | 1.000000 | 0.901235 | 0.533333 | 0.257143 | ... | 1      | 0       | 1          | 0          | 1     | 0      | 0      | 1       | 1          | 0              |
| 16  | 4.0 | 3.0      | 0.851 | 0.25      | 0.618026 | 0.562092 | 0.728395 | 0.000000 | 0.285714 | ... | 0      | 1       | 1          | 0          | 0     | 1      | 0      | 1       | 1          | 0              |
| 17  | 3.0 | 2.0      | 0.905 | 1.00      | 0.965665 | 0.522876 | 0.641975 | 0.666667 | 0.000000 | ... | 0      | 1       | 0          | 1          | 1     | 0      | 1      | 0       | 1          | 0              |
| 20  | 4.0 | 0.0      | 0.203 | 0.75      | 0.158798 | 0.196078 | 0.160494 | 0.166667 | 0.171429 | ... | 1      | 0       | 1          | 0          | 1     | 0      | 0      | 1       | 1          | 0              |
| 21  | 4.0 | 1.0      | 0.784 | 0.00      | 0.725322 | 0.313725 | 0.481481 | 0.566667 | 0.714286 | ... | 1      | 0       | 0          | 1          | 0     | 1      | 1      | 0       | 1          | 0              |
| 22  | 4.0 | 1.0      | 0.676 | 0.25      | 0.600858 | 0.104575 | 0.160494 | 0.533333 | 0.257143 | ... | 1      | 0       | 1          | 0          | 1     | 0      | 1      | 0       | 1          | 0              |
| 26  | 4.0 | 0.0      | 0.568 | 0.50      | 0.270386 | 0.843137 | 1.000000 | 0.400000 | 0.742857 | ... | 1      | 0       | 1          | 0          | 0     | 1      | 1      | 0       | 1          | 0              |


### Paste the markdown table here

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

Answer: to reduce the hallucinations from LLM when performing data conversions, it is important to specify the tasks clearly, provide some contextual information and limit the scope of content will be helpful. When asking, be precise and specific as possible and prevent to general wordings. Before prompt the tasks, provide some background knowledge so the model knows what we expect and perform better. Feeding a whole big chunk of data including unnecessary information can overwhelm the model so divide and separate the data on needs only helps to reduce the workload and potential of hallucinations. Giving it an example of the desired output is also a great way as it will generate results that strictly align to our expectation. Finally, giving feedbacks and repeat prompting in different ways will change the way the model approach the tasks and likely to prevent hallucinations gradually. 

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [22]:
# Code to rename all the columns in the dataset
abbreviation_mapping = {
    "age": "Age",
    "bp": "Blood Pressure",
    "sg": "Specific Gravity",
    "al": "Albumin",
    "su": "Sugar",
    "rbc": "Red Blood Cells",
    "pc": "Pus Cell",
    "pcc": "Pus Cell Clumps",
    "ba": "Bacteria",
    "bgr": "Blood Glucose Random",
    "bu": "Blood Urea",
    "sc": "Serum Creatinine",
    "sod": "Sodium",
    "pot": "Potassium",
    "hemo": "Hemoglobin",
    "pcv": "Packed Cell Volume",
    "wbcc": "White Blood Cell Count",
    "rbcc": "Red Blood Cell Count",
    "htn": "Hypertension",
    "dm": "Diabetes Mellitus",
    "cad": "Coronary Artery Disease",
    "appet": "Appetite",
    "pe": "Pedal Edema",
    "ane": "Anemia",
    "class": "Chronic Kidney Disease (CKD)"
}

# Rename the columns in the DataFrame
encoded_df = final_df.rename(columns=abbreviation_mapping, inplace=False)

# Display the updated DataFrame
display(encoded_df)

,Albumin,Sugar,Age,Blood Pressure,Blood Glucose Random,Blood Urea,Serum Creatinine,Sodium,Potassium,Hemoglobin,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
2,2.0,0.0,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.657143,0.172131,...,0,1,0,1,0,1,0,1,1,0
4,2.0,0.0,0.675676,0.75,0.253219,0.633987,0.777778,0.366667,0.542857,0.286885,...,1,0,1,0,1,0,1,0,1,0
7,3.0,4.0,0.851351,0.25,0.832618,0.503268,0.283951,0.333333,0.314286,0.565574,...,0,1,1,0,0,1,1,0,1,0
8,3.0,0.0,0.756757,0.25,0.223176,0.209150,0.160494,0.533333,0.514286,0.573770,...,1,0,1,0,1,0,1,0,1,0
10,4.0,0.0,0.000000,0.00,0.103004,0.372549,0.074074,0.500000,0.571429,0.352459,...,1,0,0,1,1,0,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,0.0,0.0,0.297297,0.00,0.038627,0.261438,0.012346,0.833333,0.600000,0.983607,...,1,0,1,0,1,0,1,0,0,1
149,0.0,0.0,0.229730,0.50,0.175966,0.156863,0.086420,0.833333,0.314286,0.713115,...,1,0,1,0,1,0,1,0,0,1
150,0.0,0.0,0.391892,0.00,0.150215,0.189542,0.012346,0.500000,0.285714,0.745902,...,1,0,1,0,1,0,1,0,0,1
151,0.0,0.0,0.364865,0.50,0.248927,0.183007,0.024691,0.500000,0.285714,0.614754,...,1,0,1,0,1,0,1,0,0,1


## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [24]:
# Converted SQL to Pandas code
result = final_df.groupby('Target_ckd').size().reset_index(name='count')

# Display the result
display(result)

,Target_ckd,count
0,0,111
1,1,23


The code makes a count operation on the 'Target_ckd' column, grouping the data based on whether a patient has chronic kidney disease (CKD) or not. From the result, we can see that out of the total patients in our final processed dataset, 111 patients do not have CKD (Target_ckd = 0) and 23 patients have CKD (Target_ckd = 1). This gives us a quick overview that there is less people with CKD in our cleaned dataset. In other words, this counts the occurrence of unique values in 'Target_ckd' column of the 'final_df' DataFrame and gives an understanding of distributions of cases.